Adding migration guide for the extraction of the legacy modules. (#1541)
keycloak/keycloak-documentation: upgrading/topics/keycloak/changes-19_0_0.adoc

> = Slimming the API of KeycloakSession, deprecation of federation

Reviewer: An introductory sentence would be preferrable.

```suggestion
= Changes affecting developers

{project_name} undergoes large refactoring which impacts existing code. Some of the changes
require updates to existing code. These are in more detailed described below.
```

== Why it has changed

The new map storage functionality will provide a better abstraction for the service layer and will provide a unified interface to various storages.
The previous distinction of local storage and federated storage will no longer exist, also the distinction between cached and non-cached storage disappears.

The map storage divides storage in different areas, for example users, group, clients, etc.
Each of these areas can be serviced by a different storage backend.

> With tree storage that is a functionality available within map storage, an administrator can combine several storage backends in a single area, storing different attributes and entities in different stores.

Reviewer: ```suggestion
== Rationale for changes

{project_name} has several limitations, for example downtime is needed
for upgrading a {project_name} cluster. To address the limitations, in-depth refactor has been
initiated.

The changes in this version are mostly attached to storage refactoring and a preparation
of a new storage, called map storage, which will eventually replace the current storage, which
with this version becomes called a _legacy store_.
The legacy store will still be available in {project_name} for several more versions.

The new store imposes strict separation of responsibility between service and storage layers.
For that reason, service layer visibility of objects origin will be restricted, so it will not be able to
discriminate cached or non-cached object, or objects originating in local or federated storage.

User storage SPI will become deprecated. It will be supported for several more versions, but
will be eventually replaced by map storage SPI which will offer ability to create custom
storages for any recognized area, e.g. users, roles, clients or groups.

Extensions that rely on the level of detail available to services in the legacy store will need
adjustment to retain this ability for the full deprecation period of legacy store.
How that is done is explained in the sections below.

Using legacy and map store is mutually exclusive - one cannot be used while the other is active.
```

== Changes in the module structure

As part of introducing the new storage functionality, several public APIs around storage functionality in `KeycloakSession` have been consolidated, and some have been deprecated and will be removed in one of the next versions.

> Three new modules have been introduced, and data-oriented code from server-spi, server-spi-private, and services have been moved there:

Reviewer: ```suggestion
Three new modules have been introduced, and data-oriented code from `server-spi`, `server-spi-private`, and `services` modules have been moved there:
```

After that period, they will be removed.

== Changes in `KeycloakSession`

`KeycloakSession` has been simplified.
Several methods have been deprecated in `KeycloakSession` and will be removed in a future version.

`KeycloakSession` session contains several methods for obtaining a provider for a particular object type, e.g. for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, `userStorageManager()`, and `userFederatedStorage()`.
This is confusing for the developer who has to understand the exact meaning of each method, and depends on current store layout.
The new store does not distinguish federated from local storage.

For both reasons, only the `users()` methods will be kept in `KeycloakSession` and the rest of methods are currently deprecated, return the same object as `users()`, and will eventually be removed.
The same pattern of deprecation applies to methods all other object areas, e.g. `clients()` or `groups()`.

The deprecated methods in KeycloakSession will be removed in a future release.
The `keycloak-model-legacy-*` modules will be available for a longer time, and will eventually be removed as well.

> === Migrating existing providers that don't depend on the legacy store

Reviewer: ```suggestion
=== Migrating existing providers that do not depend on the legacy store
```

Most of the existing providers need no migration.
Please review your code if it depends on any functionality in regard to cached vs. noncached or local vs. federated storage.
As an example, `userLocalStorage()` will now return the same result as `users()`, which might involve a cache if that has been enabled in the local setup.

.Before migration: accessing a deprecated API that redirects to `.users()`
[source,java,subs="+quotes"]
----
session**.userLocalStorage()**;
----

.After migration: accessing the old functionality via the LegacyStoreManagers API
[source,java,subs="+quotes"]
----
session**.users()**;
----

> 

Reviewer: The title is confusing.

Author: Actually it was a copy-paste error, will be changed in the next commit to "After migration: accessing the new API caller doesn't depend on the legacy storage API"

=== Migrating existing providers that depend on the legacy store

In the rare case when a custom provider needs to distinguish between the mode of a particular provider, access to the deprecated objects is provided by using the `LegacyStoreManagers` data store provider.
This will be available only if the legacy modules are part of the deployment.

.Before migration: accessing a deprecated API that redirects to `.users()`
[source,java,subs="+quotes"]
----
session**.userLocalStorage()**;
----

.After migration: accessing the old functionality via the LegacyStoreManagers API
[source,java,subs="+quotes"]
----
((LegacyDatastoreProvider) session.getProvider(DatastoreProvider.class))**.userLocalStorage()**;
----

Some user storage related APIs have been wrapped in `org.keycloak.storage.UserStorageUtil` for convenience.

=== Creating custom storage providers

The API for creating a custom storage provider has not been fully stabilized yet, though it is available as a tech preview.
See the `MapStorageProvider` SPI and its Javadoc for details.
The availability of the new API is a priority for the next Keycloak version.

== Interface `UserCache` moved to the legacy module

> As the caching will be trasparent with the new map store, the interface `UserCache` has been moved to the module `keycloak-legacy`.

Reviewer: ```suggestion
As the caching status of objects will be trasparent to services, the interface `UserCache`
has been moved to the module `keycloak-legacy`.
```

While such calls might be necessary while caching with the legacy store is used, it will not be necessary when using the new map store, as that one handles caching transparently.

.Before migration: code will not compile due to a changed return type
[source,java,subs="+quotes"]
----
// session.userCache() might return null, null-check omitted for brevity.
session**.userCache()**.evict(realm, user);
----

.After migration: use the API directly
[source,java,subs="+quotes"]
----
// session.getProvider(UserCache.class) might return null, null-check omitted for brevity.
session.**getProvider(UserCache.class)**.evict(realm, user);
----

To trigger the invalidation of a realm, instead of using the `UserCache` API, consider triggering an event:

.Before migration: code will not compile due to a changed return type
[source,java,subs="+quotes"]
----
UserCache cache = session.getProvider(UserCache.class);
if (cache != null) cache.clear();
----

.After migration: use the API directly
[source,java,subs="+quotes"]
----
session.invalidate(InvalidationHandler.ObjectType.REALM, realm.getId());
----

== Credential management for users

> Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)` are now managed with `user.credentialManager()._method_(\...)` which gets this functionality closer to API of users, and won't require a distinction of where the user is localed in regard to realm and storage.

Reviewer: ```suggestion
Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)`.
The new way is to leverage `user.credentialManager()._method_(\...)`. This form gets
the credential functionality closer to API of users, and does not rely on
prior knowledge of user credential location in regard to realm and storage.
```

.Before migration: code will not compile due to the new method `credentialManager()` required by `UserModel`
[source,java,subs="+quotes"]
----
public class MyUserStorageProvider implements UserLookupProvider, ... {
    /* ... */
    protected UserModel createAdapter(RealmModel realm, String username) {
        return new AbstractUserAdapter(session, realm, model) {
            @Override
            public String getUsername() {
                return username;
            }
        };
    }
}
----

.After migration: implementation of the API `UserModel.credentialManager()` for the legacy store.
[source,java,subs="+quotes"]
----
public class MyUserStorageProvider implements UserLookupProvider, ... {
    /* ... */
    protected UserModel createAdapter(RealmModel realm, String username) {
        return new AbstractUserAdapter(session, realm, model) {
            @Override
            public String getUsername() {
                return username;
            }

            @Override
            public SubjectCredentialManager credentialManager() {
                return new LegacyUserCredentialManager(session, realm, this);
            }
        };
    }
}
----

> 

Reviewer: One more item should be added there - Changes in `RealmModel`, specifically that `getUserStorageProviders` and `getUserStorageProvidersStream` methods are removed, and `UserStorageUtil.getUserStorageProvidersStream` needs to be used instead.

Author: done: 2314f72399fa8ffcc9b7c5ab2b5e051fc6c7eeee

= Changes affecting developers

> {project_name} undergoes large refactoring which impacts existing code.

Reviewer 2: ```suggestion
{project_name} undergoes large refactoring, which impacts existing code.
```

Some of these changes require updates to existing code.
These are in more detailed described below.

== Rationale for changes

{project_name} has several limitations; for example, downtime is needed for upgrading a {project_name} cluster.

> To address the limitations, in-depth refactor has been initiated.

Reviewer 2: ```suggestion
To address the limitations, an in-depth refactor has been initiated.
```

> The changes in this version are mostly attached to storage refactoring and a preparation of a new storage, called map storage, which will eventually replace the current storage, which with this version becomes called a _legacy store_.

Reviewer 2: ```suggestion
The changes in this version are mostly attached to storage refactoring and a preparation of a new storage, called map storage. This storage will eventually replace the current storage, which will be called a _legacy store_ with this version.
```

The legacy store will still be available in {project_name} for several more versions.

The new store imposes a strict separation of responsibility between the service and storage layers.
For that reason, the service layer's visibility of an object's origin will be restricted, so it will not be able to discriminate between cached or non-cached objects, or objects originating from local or federated storage.

User storage SPI will become deprecated.

> It will be supported for several more versions, but will be eventually replaced by map storage SPI which will offer the ability to create custom storages for any recognized area, e.g. users, roles, clients or groups.

Reviewer 2: ```suggestion
It will be supported for several more versions, but will be eventually replaced by the Map Storage SPI, which will offer the ability to create custom storages for any recognized area, such as users, roles, clients, or groups.
```

Extensions that rely on the level of detail available to services in the legacy store will need adjustment to retain this ability for the full deprecation period of the legacy store.

> How that is done is explained in the sections below.

Reviewer 2: ```suggestion
The following section describes how that adjustment is accomplished.
```

> Using legacy and map store is mutually exclusive: one cannot be used while the other is active.

Reviewer 2: ```suggestion
Using a legacy and map store is mutually exclusive; one store cannot be used while the other is active.
```

== Changes in the module structure

As part of introducing the new storage functionality, several public APIs around storage functionality in `KeycloakSession` have been consolidated, and some have been deprecated and will be removed in one of the next versions.
Three new modules have been introduced, and data-oriented code from `server-spi`, `server-spi-private`, and `services` modules have been moved there:

`org.keycloak:keycloak-model-legacy`::

> Contains all public facing APIs from the legacy store, like the User Storage API.

Reviewer 2: ```suggestion
Contains all public facing APIs from the legacy store, such as the User Storage API.
```

`org.keycloak:keycloak-model-legacy-private`::

> Contains private implementations that relate to user storage management (e.g. storage `*Manager` classes).

Reviewer 2: ```suggestion
Contains private implementations that relate to user storage management, such as storage `*Manager` classes.
```

`org.keycloak:keycloak-model-legacy-services`::
Contains all REST endpoints that directly operate on the legacy store, and have no meaning in the new store.

These modules will be available as long as legacy stores will be supported.
After that period, they will be removed.

== Changes in `KeycloakSession`

`KeycloakSession` has been simplified.
Several methods have been deprecated in `KeycloakSession` and will be removed in a future version.

> `KeycloakSession` session contains several methods for obtaining a provider for a particular object type, e.g. for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, `userStorageManager()`, and `userFederatedStorage()`.

Reviewer 2: ```suggestion
`KeycloakSession` session contains several methods for obtaining a provider for a particular object type, such as for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, `userStorageManager()`, and `userFederatedStorage()`.
```

> This is confusing for the developer who has to understand the exact meaning of each method, and depends on current store layout.

Reviewer 2: ```suggestion
This situaton may be confusing for the developer who has to understand the exact meaning of each method, and depends on current store layout.
```

The new store does not distinguish federated from local storage.

For both reasons, only the `users()` methods will be kept in `KeycloakSession` and the rest of methods are currently deprecated, return the same object as `users()`, and will eventually be removed.
The same pattern of deprecation applies to methods all other object areas, e.g. `clients()` or `groups()`.

The deprecated methods in KeycloakSession will be removed in a future release.

> The `keycloak-model-legacy-*` modules will be available for a longer time, and will eventually be removed as well.

Reviewer 2: ```suggestion
The `keycloak-model-legacy-*` modules will be available for a longer time, but they will eventually be removed.
```

=== Migrating existing providers that do not depend on the legacy store

Most of the existing providers need no migration.

> Please review your code if it depends on any functionality in regard to cached vs. noncached or local vs. federated storage.

Reviewer 2: ```suggestion
Please review your code if it depends on any functionality in regard to cached versus noncached or local versus federated storage.
```

As an example, `userLocalStorage()` will now return the same result as `users()`, which might involve a cache if that has been enabled in the local setup.

.Before migration: accessing a deprecated API that redirects to `.users()`
[source,java,subs="+quotes"]
----
session**.userLocalStorage()**;
----

> .After migration: accessing the new API caller doesn't depend on the legacy storage API

Reviewer: Please as a general rule, avoid shortened form.
```suggestion
.After migration: accessing the new API caller does not depend on the legacy storage API
```

[source,java,subs="+quotes"]
----
session**.users()**;
----

=== Migrating existing providers that depend on the legacy store

In the rare case when a custom provider needs to distinguish between the mode of a particular provider, access to the deprecated objects is provided by using the `LegacyStoreManagers` data store provider.

> This will be available only if the legacy modules are part of the deployment.

Reviewer 2: ```suggestion
This option will be available only if the legacy modules are part of the deployment.
```

While such calls might be necessary while caching with the legacy store is used, it will not be necessary when using the new map store, as that one handles caching transparently.

.Before migration: code will not compile due to a changed return type
[source,java,subs="+quotes"]
----
// session.userCache() might return null, null-check omitted for brevity.
session**.userCache()**.evict(realm, user);
----

.After migration: use the API directly
[source,java,subs="+quotes"]
----
// session.getProvider(UserCache.class) might return null, null-check omitted for brevity.
session.**getProvider(UserCache.class)**.evict(realm, user);
----

To trigger the invalidation of a realm, instead of using the `UserCache` API, consider triggering an event:

.Before migration: code will not compile due to a changed return type
[source,java,subs="+quotes"]
----
UserCache cache = session.getProvider(UserCache.class);
if (cache != null) cache.clear();
----

.After migration: use the invalidation API
[source,java,subs="+quotes"]
----
session.invalidate(InvalidationHandler.ObjectType.REALM, realm.getId());
----

== Credential management for users

> Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)`.

Reviewer 2: ```suggestion
Credentials for users were previously managed using `session.userCredentialManager()._method_(realm, user, \...)`.
```

= Slimming the API of KeycloakSession, deprecation of federation

== What has changed

> As part of introducing the Storage.X map storage functionality, several APIs around storage have been consolidated and some have been deprecated.

Reviewer: ```suggestion
As part of introducing the new storage functionality, several public APIs around
storage functionality in `KeycloakSession` have been consolidated, and some
have been deprecated and will be removed in one of the next versions.
```

= Slimming the API of KeycloakSession, deprecation of federation

== What has changed

As part of introducing the Storage.X map storage functionality, several APIs around storage have been consolidated and some have been deprecated.

> Several deprecated classes now exist in the modules `keycloak-model-legacy`, `keycloak-model-legacy-private` and `keycloak-model-legacy-services`.

Reviewer: ```suggestion
Three new modules have been introduced, and data-oriented code from server-spi, server-spi-private, and services has been moved there:
* `org.keycloak:keycloak-model-legacy` - contains all public facing APIs from legacy store, like User Storage API
* `org.keycloak:keycloak-model-legacy-private` - contains private implementations that relates to user storage management (e.g. storage `*Manager` classes).
* `org.keycloak:keycloak-model-legacy-services` - contains all REST endpoints that directly operate on legacy store, and have no meaning in the new store

These modules will be available as long as legacy store would be supported. After that period, they will be removed.
```

> Also, several methods on KeycloakSession have been deprecated.

Reviewer: ```suggestion

== Changes in `KeycloakSession`

`KeycloakSession` has been simplified. Several methods have been deprecated in `KeycloakSession` and will be removed in future version.

`KeycloakSession` session contains several methods for obtaining a provider for a particular object
type, e.g. for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, 
`userStorageManager()`, `userFederatedStorage()`. This is confusing for the developer who has to
exactly understand the meaning of each of the methods, and depends on current store layout. The new store does not distinguish federated from local storage.

For both reasons, only `users()` methods will be kept in `KeycloakSession` and the rest of methods
are currently deprecated and return the same object as `users()`, and will eventually be removed.
The same pattern of deprecation applies to methods all other object areas, e.g. `clients()` or `groups()`.

=== Migrating existing providers requiring legacy store

Most of existing providers need no migration. In the rare case when a custom provider needs to
distinguish between the mode of a particular provider, access to the original objects deprecated
is provided by using the `LegacyStoreManagers` data store provider.

<< Show some old code accessing say userLocalStorage and its translation to new code >>

=== Creating custom storage providers

The API for creating a custom storage provider has not been fully stabilized yet, 
though it is available as a tech preview. See `MapStorageProvider` SPI and Javadoc there.
The availability of the new API is a priority for next Keycloak version.
```

== Why it has changed

The new Storage.X map storage functionality will provide a better abstraction for the service layer and will provide a unified interface to various storages.
The previous distinction of local storage and federated storage will no longer exist, also the distinction between cached and non-cached storage disappears.

The map storage divides storage in different areas, for example users, group, clients, etc.
Each of this area can be serviced by a different storage backend.

> With tree storage that is a functionality available within map storage, an administrator can combine several storage backends in a single area, storing different attributes and entities in different stores.

Reviewer: Apart from _Storage.X_ reference, I like this paragraph.

Author: Storage.X has been removed.

> == API Changes

Reviewer: This section should give also code samples of original code and how this code should be rewritten for each of the paragraphs, and that for both legacy store (using `LegacyStoreManagers`) and for new store (e.g. just using `users()` instead of `userLocalStorage()`).

It should also be stated clearly that using `LegacyStoreManagers` will be only supported for the legacy store and not in the new store (regardless it is obvious from the name).

Author: I've added the code examples in each paragraph instead in in its own API block to keep things together.
